packages/ui-kit/src/components/Counter/Counter.tsx

```diff
@@ -1,135 +1,136 @@
-import { css } from '@emotion/css'
+import classnames from 'classnames'
 import React from 'react'
-import { getTimeZone, PROPEL_GRAPHQL_API_ENDPOINT, useCounterQuery } from '../../helpers'
-import type { ChartStyles } from '../../themes'
-import { defaultStyles } from '../../themes'
+import { getTimeZone, PROPEL_GRAPHQL_API_ENDPOINT, useCombinedRefsCallback, useCounterQuery } from '../../helpers'
 import { ErrorFallback } from '../ErrorFallback'
 import { Loader } from '../Loader'
+import { useTheme } from '../ThemeProvider'
 import { withContainer } from '../withContainer'
+import componentStyles from './Counter.module.scss'
 import type { CounterProps } from './Counter.types'
 import { getValueWithPrefixAndSufix } from './utils'
 
-export const CounterComponent = (props: CounterProps) => {
-  const {
-    value: staticValue,
-    query,
-    prefixValue,
-    sufixValue,
-    styles,
-    loading: isLoadingStatic = false,
-    localize,
-    ...rest
-  } = props
-
-  /**
-   * If the user passes `value` attribute, it
-   * should behave as a static component without any GraphQL operation performed
-   */
-  const isStatic = !query
-
-  const [propsMismatch, setPropsMismatch] = React.useState(false)
-
-  const counterRef = React.useRef<HTMLSpanElement>(null)
-
-  const {
-    isInitialLoading: isLoadingQuery,
-    error,
-    data: fetchedValue
-  } = useCounterQuery(
+export const CounterComponent = React.forwardRef<HTMLSpanElement, CounterProps>(
+  (
     {
-      endpoint: query?.propelApiUrl ?? PROPEL_GRAPHQL_API_ENDPOINT,
-      fetchParams: {
-        headers: {
-          'content-type': 'application/graphql-response+json',
-          authorization: `Bearer ${query?.accessToken}`
-        }
-      }
+      value: staticValue,
+      query,
+      prefixValue,
+      sufixValue,
+      loading: isLoadingStatic = false,
+      localize,
+      className,
+      baseTheme,
+      loaderProps,
+      errorFallbackProps,
+      timeZone,
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      card,
+      ...rest
     },
-    {
-      counterInput: {
-        metricName: query?.metric,
-        timeZone: props.timeZone ?? getTimeZone(),
-        timeRange: {
-          relative: query?.timeRange?.relative ?? null,
-          n: query?.timeRange?.n ?? null,
-          start: query?.timeRange?.start ?? null,
-          stop: query?.timeRange?.stop ?? null
-        },
-        filters: query?.filters ?? []
+    forwardedRef
+  ) => {
+    const innerRef = React.useRef<HTMLSpanElement>(null)
+    const { componentContainer, setRef } = useCombinedRefsCallback({ innerRef, forwardedRef })
+    useTheme({ componentContainer, baseTheme })
```

Comment:
Eventually, we'll want to tell customers how to construct custom, themeable components themselves. Is this going to be the suggested way to do that?

1. Accept a `baseTheme?: DefaultThemes` in your custom component's props.
2. Construct an `innerRef`.
3. Invoked `useCombinedRefsCallback`.
4. Pass this to `useTheme`.

Is that correct? Basically the code snippet above?

I'd ask, is there some way to simplify this? Could we/should we merge some of the setup into `useTheme` for public consumption?

Reply:
Yeah, probably I could merge `useCombinedRefsCallback` and `useTheme`. I suggest to do that in the following updates.